### DISCUSSION TITLE
:checkhealth rustaceanvim says rust-analyzer is found, but isn't used

### DISCUSSION THREAD

**Discussion by FritsHoogland** [mrcjkb/rustaceanvim#164] [Q&A] [answered]
I am running rustaceanvim using nvim with astronvim, all latest version.
`:checkhealth rustaceanvim` says rust-analyzer is found, but the rustaceanvim plugin does not work.
Once I add rust-analyzer using `rustup component add rust-analyzer`, it all works.
Is that correct behaviour?

**Comment by mrcjkb** [comment, marked as answer]
My guess is that the `rust-analyzer` binary that is detected by the health check is broken/fails to start, and the one installed using rustup isn't.

You should be able to troubleshoot this by running it from the command line.

UPDATE: rustup comes with the rust-analyzer wrapper, which is what `checkhealth` picks up.
But the wrapper will fail to start if you don't `rustup component add rust-analyzer`.

Created https://github.com/mrcjkb/rustaceanvim/issues/284

**Reply by FritsHoogland** [reply]
This is probably a feature request, but thinking about your answer:
Wouldn't it be great of check health would additional to identifying rust-analyzer, doing a simple check to see if it can use it prevent a lot of doubt? I was very confused to find that the health check result didn't mean the plugin could use rust-analyzer, but just that it found it.

**Reply by mrcjkb** [reply]
🤔 the only sure way to tell if rust-analyzer isn't broken as an LSP server is to start an LSP client in a rust file or project 
Even then, the failure to start could be for various reasons (e.g. a misconfigured project).

I could potentially report on previous failed LSP client startups in the health check, but I'm not sure if it adds any value.
It would probably be better for the Neovim LSP client to report an error if the server fails to start.

A broken rust-analyzer could be starting successfully, but sending empty responses to the LSP client. In such a case, it would be very difficult for the client to tell that something is wrong.

**Reply by FritsHoogland** [reply]
I agree, and sorry to sound so demanding.
Currently the check simply detects if it can find rust-analyzer, right? 

My suggestion of performing a check by running something through the found rust-analyzer is based on the assumption that the 'faulty' rust-analyzer found will throw an error for anything it is asked to do. I do not know if that is true.

If it's true that the found rust-analyzer before "rustup component add rust-analyzer" errors on everything, then I think adding a check by requesting something simple via it makes sense. If the assumption is not true, and it's much more nuanced and it cannot be known if all or certain things will work or will not work, then adding a test doesn't really mean anything, and thus cannot be done as a definitive fact for the health check, and therefore should not be added.

**Reply by mrcjkb** [reply]
> I agree, and sorry to sound so demanding.

No worries :smile:

> Currently the check simply detects if it can find rust-analyzer, right?

It tries to run `rust-analyzer --version`.
